File: clientFunctions/java-clientFunctions/src/main/java/com/example/ApiClient.java
package com.example;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public class ApiClient {

    private final String baseUrl;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public ApiClient(String baseUrl) {
        this.baseUrl = baseUrl;
        this.httpClient = HttpClient.newHttpClient();
        this.objectMapper = new ObjectMapper();
    }

    private JsonNode sendRequest(HttpRequest request) {
        try {
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() == 200) {
                return objectMapper.readTree(response.body());
            } else {
                System.err.println("HTTP request failed with code: " + response.statusCode());
                return null;
            }
        } catch (IOException | InterruptedException e) {
            System.err.println("Error during HTTP request: " + e.getMessage());
            return null;
        }
    }

    public JsonNode execQuery(String query) {
        String jsonData = String.format("{\"query\": \"%s\"}", query);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/execQuery"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
        return sendRequest(request);
    }

    public boolean createTable(String tableSql) {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/createTable"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(tableSql))
                .build();
        JsonNode response = sendRequest(request);
        return response != null && "success".equals(response.path("status").asText());
    }

    public JsonNode listTables() {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/listTables"))
                .GET()
                .build();
        return sendRequest(request);
    }

    public boolean insert(String tableName, String values) {
        try {
            // Validate that the input "values" is a JSON object
            ObjectMapper objectMapper = new ObjectMapper();
            objectMapper.readTree(values); // Ensures it's valid JSON

            // Create the JSON payload for the HTTP request
            String jsonData = String.format("{\"tableName\": \"%s\", \"values\": %s}", tableName, values);

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + "/api/insert"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                    .build();

            // Send the request and process the response
            JsonNode response = sendRequest(request);
            return response != null && "success".equals(response.path("status").asText());
        } catch (Exception e) {
            System.err.println("Error during insert: " + e.getMessage());
            return false;
        }
    }


    public boolean deleteData(String tableName, String columns, String values) {
        String jsonData = String.format("{\"tableName\": \"%s\", \"columns\": %s, \"values\": %s}", tableName, columns, values);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/delete"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
        JsonNode response = sendRequest(request);
        return response != null && "success".equals(response.path("status").asText());
    }

    public JsonNode select(String tableName, String columns, String whereClause, String params) {
        String jsonData = String.format(
            "{\"tableName\": \"%s\", \"columns\": %s, \"whereClause\": \"%s\", \"params\": %s}",
            tableName, columns, whereClause, params
        );
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/select"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonData))
                .build();
        return sendRequest(request);
    }
}

File: clientFunctions/c++-clientFunctions/ApiClient.cpp
#include "ApiClient.h"
#include <curl/curl.h>
#include <iostream>

ApiClient::ApiClient(const std::string& baseUrl) : baseUrl(baseUrl) {}

ApiClient::~ApiClient() {}

// Callback function for libcurl to write response data into a string
static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userData) {
    size_t totalSize = size * nmemb;
    userData->append(static_cast<char*>(contents), totalSize);
    return totalSize;
}

std::string ApiClient::sendRequest(const std::string& url, const std::string& method, const std::string& body) {
    CURL* curl = curl_easy_init();
    if (!curl) {
        std::cerr << "Failed to initialize libcurl" << std::endl;
        return "";
    }

    std::string responseData;
    struct curl_slist* headers = nullptr;

    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &responseData);

    // Set headers
    headers = curl_slist_append(headers, "Content-Type: application/json");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

    if (method == "POST") {
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
    }

    CURLcode res = curl_easy_perform(curl);
    if (res != CURLE_OK) {
        std::cerr << "libcurl error: " << curl_easy_strerror(res) << std::endl;
    }

    curl_easy_cleanup(curl);
    curl_slist_free_all(headers);

    return responseData;
}

nlohmann::json ApiClient::execQuery(const std::string& query) {
    std::string jsonData = "{\"query\": \"" + query + "\"}";
    std::string response = sendRequest(baseUrl + "/api/execQuery", "POST", jsonData);
    return response.empty() ? nullptr : nlohmann::json::parse(response);
}

bool ApiClient::createTable(const std::string& tableSql) {
    std::string response = sendRequest(baseUrl + "/api/createTable", "POST", tableSql);
    if (response.empty()) return false;
    auto jsonResponse = nlohmann::json::parse(response);
    return jsonResponse["status"] == "success";
}

nlohmann::json ApiClient::listTables() {
    std::string response = sendRequest(baseUrl + "/api/listTables", "GET");
    return response.empty() ? nullptr : nlohmann::json::parse(response);
}

bool ApiClient::insert(const std::string& tableName, const std::string& values) {
    try {
        nlohmann::json::parse(values); // Validate JSON
        std::string jsonData = "{\"tableName\": \"" + tableName + "\", \"values\": " + values + "}";
        std::string response = sendRequest(baseUrl + "/api/insert", "POST", jsonData);
        if (response.empty()) return false;
        auto jsonResponse = nlohmann::json::parse(response);
        return jsonResponse["status"] == "success";
    } catch (const std::exception& e) {
        std::cerr << "Error during insert: " << e.what() << std::endl;
        return false;
    }
}

bool ApiClient::deleteEntry(const std::string& tableName, const std::string& columns, const std::string& values) {
    std::string jsonData = "{\"tableName\": \"" + tableName + "\", \"columns\": " + columns + ", \"values\": " + values + "}";
    std::string response = sendRequest(baseUrl + "/api/delete", "POST", jsonData);
    if (response.empty()) return false;
    auto jsonResponse = nlohmann::json::parse(response);
    return jsonResponse["status"] == "success";
}

nlohmann::json ApiClient::select(const std::string& tableName, const std::string& columns, const std::string& whereClause, const std::string& params) {
    std::string jsonData = "{\"tableName\": \"" + tableName + "\", \"columns\": " + columns + ", \"whereClause\": \"" + whereClause + "\", \"params\": " + params + "}";
    std::string response = sendRequest(baseUrl + "/api/select", "POST", jsonData);
    return response.empty() ? nullptr : nlohmann::json::parse(response);
}

